libecole/src/instance/combinatorial-auction.cpp

```diff
@@ -0,0 +1,348 @@
+#include <fmt/format.h>
+#include <map>
+#include <xtensor/xadapt.hpp>
+#include <xtensor/xindex_view.hpp>
+#include <xtensor/xrandom.hpp>
+#include <xtensor/xsort.hpp>
+#include <xtensor/xtensor.hpp>
+#include <xtensor/xview.hpp>
+
+#include "ecole/instance/combinatorial-auction.hpp"
+#include "ecole/scip/cons.hpp"
+#include "ecole/scip/model.hpp"
+#include "ecole/scip/utils.hpp"
+#include "ecole/scip/var.hpp"
+
+namespace ecole::instance {
+
+/**************************************************
+ *  CombinatorialAuctionGenerator methods  *
+ **************************************************/
+
+CombinatorialAuctionGenerator::CombinatorialAuctionGenerator(RandomEngine random_engine_, Parameters parameters_) :
+	random_engine{random_engine_}, parameters{parameters_} {}
+CombinatorialAuctionGenerator::CombinatorialAuctionGenerator(Parameters parameters_) :
+	CombinatorialAuctionGenerator{ecole::spawn_random_engine(), parameters_} {}
+CombinatorialAuctionGenerator::CombinatorialAuctionGenerator() : CombinatorialAuctionGenerator(Parameters{}) {}
+
+scip::Model CombinatorialAuctionGenerator::next() {
+	return generate_instance(random_engine, parameters);
+}
+
+void CombinatorialAuctionGenerator::seed(Seed seed) {
+	random_engine.seed(seed);
+}
+
+namespace {
+
+using std::size_t;
+using vector = std::vector<size_t>;
+using xvector = xt::xtensor<double, 1>;
+using xvector_size_t = xt::xtensor<size_t, 1>;
+using xmatrix = xt::xtensor<double, 2>;
+
+/** Sample with replacement based on weights given by a probability or weight vector.
+ *
+ * Samples n_samples values from a weighted distribution defined by the weights.
+ * The values are in the range of [1, weights.size()].
+ */
+auto arg_choice_without_replacement(size_t n_samples, xvector weights, RandomEngine& random_engine) -> xvector_size_t {
```

Comment:
It looks to me like this is arg_choice **with** replacement no? Items can be sampled multiple time here.

I have made an [implementation](https://github.com/xtensor-stack/xtensor/blob/41fa9ce51e3d7708b3db28c97de9cad82c0f06ba/include/xtensor/xrandom.hpp#L847) in xtensor. It is not yet released but you can take inspiration.

Otherwise, if your weights are double, you can use directly `std::discrete_distribution`